pkg/resources/management.cattle.io/v3/globalrole/validator.go

```diff
@@ -124,6 +139,30 @@ func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResp
 	return admission.ResponseAllowed(), nil
 }
 
+// validateDelete checks if a global role can be deleted and returns the appropriate response.
+func validateDelete(oldRole *v3.GlobalRole, fldPath *field.Path) (*admissionv1.AdmissionResponse, error) {
+	if oldRole.Builtin {
+		return admission.ResponseBadRequest(field.Forbidden(fldPath, "cannot delete builtin GlobalRoles").Error()), nil
+	}
+	return admission.ResponseAllowed(), nil
+}
+
+// validateCreateFields blocks the creation of builtin globalRoles
+func validateCreateFields(oldRole *v3.GlobalRole, fldPath *field.Path) *field.Error {
```

Comment:
When are built-in roles created and by who? How is it bypassing this check?

Reply:
Rancher creates the built-in GlobalRoles (as well as built-in role templates): https://github.com/rancher/rancher/blob/49e4618602de246cd51188d9338fb850ffac975e/pkg/data/management/rolebuilder.go#L318. To bypass the webhook, it uses the [bypass method](https://github.com/rancher/rancher-docs/blob/08ec8f2bddb89617c8f7cd0277f7649f8bbfb61a/docs/reference-guides/rancher-webhook.md#bypassing-the-webhook) (docs are being worked on), as you can see here: https://github.com/rancher/rancher/blob/49e4618602de246cd51188d9338fb850ffac975e/pkg/data/management/rolebuilder.go#L337.

Reply:
Good to know